### DISCUSSION TITLE
Within a custom model, how to Expand to a custom model ?

### DISCUSSION THREAD

**Discussion by SkYNewZ** [pocketbase/pocketbase#2849] [Q&A] [answered]
Imagine I have this:

```go
type Article struct {
    models.BaseModel

    User *User  `db:"user" json:"user"`
}

type User struct {
    models.BaseModel

    Foo string  `db:"foo" json:"foo"`
}

```

If using [this guide](https://pocketbase.io/docs/custom-models/#fetch-articles) to get my Article, how can I Expand the user within the Article ?

Currently I just don't use custom models because of this and I have made a custom function to cast. This works perfectly but I found it a little bit messy.

```go
type channelSynchronisationRepository struct {
	q              *daos.Dao
	collectionName string
}

func (c *channelSynchronisationRepository) FindByChannelID(channelID string) (*ChannelSynchronisation, error) {
	record, err := c.q.FindFirstRecordByData(c.collectionName, "channel_id", channelID)
	if err != nil {
		return nil, fmt.Errorf("could not find channel synchronisation by channel ID: %w", err)
	}

	if err := c.expandUser(record); err != nil {
		return nil, err
	}

	return channelSynchronisationRecordToChannelSynchronisation(record), nil
}

func (c *channelSynchronisationRepository) expandUser(record *models.Record) error {
	errs := c.q.ExpandRecord(record, []string{"user"}, func(userCollection *models.Collection, usersIDs []string) ([]*models.Record, error) {
		return c.q.FindRecordsByIds("users", usersIDs)
	})

	if len(errs) > 0 {
		return fmt.Errorf("could not expand channel synchronisation: %v", errs)
	}

	return nil
}

func channelSynchronisationRecordToChannelSynchronisation(record *models.Record) *ChannelSynchronisation {
	user := new(User)
	user.FromRecord(record.Expand()["user"].(*models.Record))

	return &ChannelSynchronisation{
		ChannelID:  record.GetString("channel_id"),
		PlaylistID: record.GetString("playlist_id"),
		User:       user,
	}
}

func (u *User) FromRecord(record *models.Record) {
	u.BaseModel = record.BaseModel
	u.Email = record.GetString("email")
	u.Name = types.Pointer(record.GetString("name"))
	u.AvatarURL = types.Pointer(record.GetString("avatar_url"))
	u.Language = types.Pointer(record.GetString("language"))

	token := &oauth2.Token{}
	if err := record.UnmarshalJSONField("token", token); err != nil {
		log.Panicln("could not unmarshal token:", err)
	}

	u.Token = token
}

```

Do you have a better solution ?
Thanks.

**Comment by ganigeorgiev** [comment, marked as answer]
At the moment there isn't an easy way to implement lazy `expand` for custom model structs (_other than making an individual query for the wanted expand and assigning it to the struct_). The custom model struct is an alternative mechanism to query and operate with the DB. Reach to it only if you think that it'll benefit your use case, otherwise you may be better working with plain records.

A better approach for type-safe record models will be eventually implemented in the future based on https://github.com/pocketbase/pocketbase/discussions/2697#discussioncomment-6154067, but for now remains a low priority.

**Reply by SkYNewZ** [reply]
Thanks for you quick answer!

Meanwhile what do you think of my current implementation?

**Reply by ganigeorgiev** [reply]
Alternatively, if you want to populate the `User` with the Article query, you can also check the example in https://github.com/pocketbase/dbx#executing-queries, but note that this will require adding an extra `AndSelect(...)` statement to your query prefixed with the `User` field db tag (eg. `user.id` to scan into `Article.User.Id`)

**Reply by ganigeorgiev** [reply]
> Meanwhile what do you think of my current implementation?

I don't have the full picture of your use case and don't understand what is the desired access behavior or outcome, and don't feel confident enough to comment on the proposed code sample.
